Respect core.filemode when comparing filemodes on checkout

tiennou opened the pull request:
Fixes #4504

tiennou commented:
No experience with the checkout code (or the Windows permission model for that matter), but I'm suspicious of the `return false`. Shouldn't we at least check for read/write ?

tiennou commented on src/checkout.c:
```diff
@@ -70,6 +70,7 @@ typedef struct {
 	git_buf tmp;
 	unsigned int strategy;
 	int can_symlink;
+	int respect_filemode;
```
It's inspired by `can_symlink`, which comes from config too. Is `bool` generally preferred ? I can switch them in that case, or just the new.

tiennou replied:
A few remarks on that : it all comes down to `git__parse_bool` taking an `int *` as its argument. I've tried to fix the signature, but causes `bool` to appear in `config.h`, and I can't seem to find any public usage of `bool`/`stdbool.h` (our only reference is in cc-compat). So meh, but if you want a PR for that, I have the code.

ethomson replied:
Nah, this is legit.  Sorry for leading you down a rather pointless rabbit hole.

tiennou commented:
I've extracted the relevant filemode comparison function in `utils.h` and added tests tests for combinations of filemodes. Note that I dumped tests in `futils.c` which is not quite the same namespace (suggestions welcome if a better home is warranted).

I'm WIPing to see how does AppVeyor react to using that code on Windows. If that works, I'll just `#ifdef` out the strict part, because it makes no sense there 🤞.

tiennou commented:
@ethomson The POSIX part is good to go, I'll defer to you on what to do to make Windows work, because I'm just headlessly hacking around in that last commit 😐.

I suspect the not-added-in-this-PR failing tests are because we're trying to do the full `core.filemode` check but I'm not entirely sure…

tiennou commented:
I got the filemode test to work everywhere (phew), but I'm now getting checkout failures as well. So either my test is completely wrong (Travis 👀😉) ~~those are running in a repo with an incorrect `core.filemode` value (which should have been handled by `git_repository_reinit_filesystem`)~~ (tests are never strict on Windows so `core.filemode` doesn't matter there) *or* we were wrong on Windows ??

Failing tests are [1](https://github.com/libgit2/libgit2/blob/master/tests/checkout/icase.c?utf8=%E2%9C%93#L269), [2](https://github.com/libgit2/libgit2/blob/master/tests/checkout/tree.c?utf8=%E2%9C%93#L379), [3](https://github.com/libgit2/libgit2/blob/master/tests/checkout/typechange.c?utf8=%E2%9C%93#L155).

tiennou commented on src/util.h:
```diff
@@ -539,4 +538,36 @@ GIT_INLINE(double) git__timer(void)
 extern int git__getenv(git_buf *out, const char *name);
 
+#ifdef WIN32
+#define S_ISREG(m) (((m) & _S_IFMT) == _S_IFREG)
```
This is only temporary, but I didn't know where to put it (and VS2010 actually doesn't like it being redefined).

ethomson replied:
Probably safer to:
```
#ifndef S_ISREG
# define S_ISREG...
#endif
```

tiennou replied:
Actually, I wanted to move it in our compat `posix` headers for Windows (with the define guard), but I'm not sure where/which one I should use.

tiennou commented on src/util.h:
```diff
@@ -539,4 +538,36 @@ GIT_INLINE(double) git__timer(void)
+
+GIT_INLINE(bool) git__is_filemode_equal(int strict, git_filemode_t a, git_filemode_t b)
+{
+#ifdef WIN32
```
Unused `strict` parameter when WIN32-ing

ethomson commented on src/util.h:
```diff
@@ -539,4 +538,36 @@ GIT_INLINE(double) git__timer(void)
+#endif
+
+	/* we only care about rw differences */
+	return (!!(a & rw) == !!(b & rw));
```
I don't think there's a need to set up this `rw` mask, or test with it at all.  We know they're both files, you can simply `return true` here.

It's reasonable for a file to have _no_ read or write bits set, but this is still a file.  And if you can read the file despite those permissions (eg you're running as root) then this will incorrectly treat those file modes as different.

ethomson replied:
That may actually be why the Windows build is failing.  I don't have a Windows machine in front of me, but it's possible that it's reading the lack of posix permissions as `0100000` for some file(s).

ethomson commented:
I pulled this down and spent some time to figure out why it was failing on Windows.  I did some refactoring and ended up with #4545.  There were some changes to support Win32 and POSIX both; I would love your feedback.
